File: python-app/stripe_helper.py
"""
Stripe Helper Module
Handles all Stripe payment operations
"""

import json
import time
import hmac
import hashlib
import urllib.request
import urllib.parse
import urllib.error
from products import get_plan_by_price_id


class StripeHelper:
    def __init__(self, secret_key, webhook_secret=None):
        self.secret_key = secret_key
        self.webhook_secret = webhook_secret
        self.api_base = 'https://api.stripe.com/v1'

    def _request(self, method, endpoint, data=None):
        """Make API request to Stripe"""
        url = self.api_base + endpoint

        headers = {
            'Authorization': f'Bearer {self.secret_key}',
            'Content-Type': 'application/x-www-form-urlencoded'
        }

        if data:
            data = urllib.parse.urlencode(data).encode('utf-8')

        req = urllib.request.Request(url, data=data, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req) as response:
                return json.loads(response.read().decode('utf-8'))
        except urllib.error.HTTPError as e:
            error_body = json.loads(e.read().decode('utf-8'))
            raise Exception(error_body.get('error', {}).get('message', 'Stripe API error'))

    def get_or_create_customer(self, team, db):
        """Create or get Stripe customer"""
        if team.get('stripe_customer_id'):
            return team['stripe_customer_id']

        # Create new customer
        customer = self._request('POST', '/customers', {
            'metadata[team_id]': team['id']
        })

        # Update team with customer ID
        cursor = db.cursor()
        cursor.execute(
            "UPDATE teams SET stripe_customer_id = %s WHERE id = %s",
            (customer['id'], team['id'])
        )
        db.commit()
        cursor.close()

        return customer['id']

    def create_checkout_session(self, team, price_id, app_url, db):
        """Create Stripe checkout session"""
        customer_id = self.get_or_create_customer(team, db)

        session = self._request('POST', '/checkout/sessions', {
            'customer': customer_id,
            'payment_method_types[]': 'card',
            'line_items[0][price]': price_id,
            'line_items[0][quantity]': '1',
            'mode': 'subscription',
            'success_url': f'{app_url}/dashboard?success=true',
            'cancel_url': f'{app_url}/pricing?canceled=true'
        })

        return session

    def create_portal_session(self, customer_id, app_url):
        """Create billing portal session"""
        session = self._request('POST', '/billing_portal/sessions', {
            'customer': customer_id,
            'return_url': f'{app_url}/dashboard'
        })
        return session

    def handle_webhook(self, payload, sig_header, db):
        """Handle Stripe webhook"""
        if self.webhook_secret:
            self._verify_webhook_signature(payload, sig_header)

        event = json.loads(payload)
        event_type = event.get('type')
        data_object = event.get('data', {}).get('object', {})

        if event_type == 'checkout.session.completed':
            self._handle_checkout_completed(data_object, db)
        elif event_type in ['customer.subscription.updated', 'customer.subscription.deleted']:
            self._handle_subscription_change(data_object, db)

        return True

    def _verify_webhook_signature(self, payload, sig_header):
        """Verify webhook signature"""
        parts = sig_header.split(',')
        timestamp = None
        signatures = []

        for part in parts:
            key, value = part.strip().split('=', 1)
            if key == 't':
                timestamp = value
            elif key == 'v1':
                signatures.append(value)

        if not timestamp or not signatures:
            raise Exception('Invalid signature header')

        if abs(time.time() - int(timestamp)) > 300:
            raise Exception('Timestamp outside tolerance')

        signed_payload = f'{timestamp}.{payload}'
        expected = hmac.new(
            self.webhook_secret.encode(),
            signed_payload.encode(),
            hashlib.sha256
        ).hexdigest()

        if not any(hmac.compare_digest(expected, sig) for sig in signatures):
            raise Exception('Invalid signature')

    def _handle_checkout_completed(self, session, db):
        """Handle checkout completed"""
        customer_id = session.get('customer')
        subscription_id = session.get('subscription')

        if not subscription_id:
            return

        cursor = db.cursor(dictionary=True)
        cursor.execute("SELECT id FROM teams WHERE stripe_customer_id = %s", (customer_id,))
        team = cursor.fetchone()

        if team:
            subscription = self._request('GET', f'/subscriptions/{subscription_id}')
            price_id = subscription['items']['data'][0]['price']['id']
            plan = get_plan_by_price_id(price_id)

            cursor.execute("""
                UPDATE teams SET
                    stripe_subscription_id = %s,
                    stripe_product_id = %s,
                    plan_name = %s,
                    subscription_status = %s
                WHERE id = %s
            """, (
                subscription_id,
                subscription['items']['data'][0]['price'].get('product'),
                plan['name'] if plan else 'Unknown',
                subscription['status'],
                team['id']
            ))
            db.commit()

        cursor.close()

    def _handle_subscription_change(self, subscription, db):
        """Handle subscription change"""
        customer_id = subscription.get('customer')

        cursor = db.cursor(dictionary=True)
        cursor.execute("SELECT id FROM teams WHERE stripe_customer_id = %s", (customer_id,))
        team = cursor.fetchone()

        if team:
            price_id = subscription['items']['data'][0]['price']['id']
            plan = get_plan_by_price_id(price_id)

            cursor.execute("""
                UPDATE teams SET
                    stripe_subscription_id = %s,
                    stripe_product_id = %s,
                    plan_name = %s,
                    subscription_status = %s
                WHERE id = %s
            """, (
                subscription['id'],
                subscription['items']['data'][0]['price'].get('product'),
                plan['name'] if plan else 'Unknown',
                subscription['status'],
                team['id']
            ))
            db.commit()

        cursor.close()

File: ruby-app/stripe_helper.rb
# Stripe Helper for Ruby/Sinatra
# Handles Stripe API calls using net/http (no external gem required)

require 'net/http'
require 'uri'
require 'json'
require 'openssl'

class StripeHelper
  STRIPE_API_BASE = 'https://api.stripe.com/v1'.freeze

  def initialize
    @secret_key = ENV['STRIPE_SECRET_KEY'] || ''
    @webhook_secret = ENV['STRIPE_WEBHOOK_SECRET'] || ''
  end

  def create_checkout_session(team_id:, price_id:, success_url:, cancel_url:)
    return { 'error' => 'Stripe is not configured' } if @secret_key.empty?

    params = {
      'payment_method_types[]' => 'card',
      'line_items[0][price]' => price_id,
      'line_items[0][quantity]' => '1',
      'mode' => 'subscription',
      'success_url' => success_url,
      'cancel_url' => cancel_url,
      'client_reference_id' => team_id.to_s
    }

    response = api_request('POST', '/checkout/sessions', params)

    if response['error']
      { 'error' => response['error']['message'] || 'Failed to create checkout session' }
    else
      { 'sessionId' => response['id'], 'url' => response['url'] }
    end
  end

  def create_portal_session(customer_id:, return_url:)
    return { 'error' => 'Stripe is not configured' } if @secret_key.empty?

    params = {
      'customer' => customer_id,
      'return_url' => return_url
    }

    response = api_request('POST', '/billing_portal/sessions', params)

    if response['error']
      { 'error' => response['error']['message'] || 'Failed to create portal session' }
    else
      { 'url' => response['url'] }
    end
  end

  def retrieve_subscription(subscription_id)
    return nil if @secret_key.empty?
    api_request('GET', "/subscriptions/#{subscription_id}")
  end

  def cancel_subscription(subscription_id)
    return { 'error' => 'Stripe is not configured' } if @secret_key.empty?
    api_request('DELETE', "/subscriptions/#{subscription_id}")
  end

  def verify_webhook(payload, signature)
    return nil if @webhook_secret.empty? || signature.nil?

    elements = {}
    signature.split(',').each do |part|
      key, value = part.split('=', 2)
      elements[key] = value
    end

    timestamp = elements['t']
    expected_sig = elements['v1']

    return nil unless timestamp && expected_sig

    # Check timestamp is within tolerance (5 minutes)
    if (Time.now.to_i - timestamp.to_i).abs > 300
      return nil
    end

    # Compute expected signature
    signed_payload = "#{timestamp}.#{payload}"
    computed_sig = OpenSSL::HMAC.hexdigest('SHA256', @webhook_secret, signed_payload)

    # Secure compare
    return nil unless secure_compare(computed_sig, expected_sig)

    JSON.parse(payload)
  rescue JSON::ParserError
    nil
  end

  private

  def api_request(method, endpoint, params = nil)
    uri = URI.parse("#{STRIPE_API_BASE}#{endpoint}")

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.verify_mode = OpenSSL::SSL::VERIFY_PEER

    case method
    when 'GET'
      request = Net::HTTP::Get.new(uri.request_uri)
    when 'POST'
      request = Net::HTTP::Post.new(uri.request_uri)
      request.set_form_data(params) if params
    when 'DELETE'
      request = Net::HTTP::Delete.new(uri.request_uri)
    end

    request['Authorization'] = "Bearer #{@secret_key}"
    request['Content-Type'] = 'application/x-www-form-urlencoded'

    response = http.request(request)
    JSON.parse(response.body)
  rescue StandardError => e
    { 'error' => { 'message' => e.message } }
  end

  def secure_compare(a, b)
    return false if a.bytesize != b.bytesize
    l = a.unpack("C*")
    r = b.unpack("C*")
    result = 0
    l.zip(r) { |x, y| result |= x ^ y }
    result == 0
  end
end
